src/gazebo_mavlink_interface.cpp

```diff
@@ -417,30 +432,77 @@ void GazeboMavlinkInterface::handle_actuator_controls(const gz::sim::UpdateInfo
 
   last_actuator_time_ = _info.simTime;
 
-  for (unsigned i = 0; i < n_out_max; i++) {
-    input_index_[i] = i;
-  }
-  // Read Input References
-  input_reference_.resize(n_out_max);
-
   Eigen::VectorXd actuator_controls = mavlink_interface_->GetActuatorControls();
   if (actuator_controls.size() < n_out_max) return; //TODO: Handle this properly
 
-  for (int i = 0; i < input_reference_.size(); i++) {
+  // Read Cmd vel input for rover
+  if (actuator_controls[n_out_max - 1] != 0.0 || actuator_controls[n_out_max - 2] != 0.0) {
+    cmd_vel_thrust_ = armed ? actuator_controls[n_out_max - 1] : 0.0;
+    cmd_vel_torque_ = armed ? actuator_controls[n_out_max - 2] : 0.0;
+    input_is_cmd_vel_ = true;
+    received_first_actuator_ = mavlink_interface_->GetReceivedFirstActuator();
+    return;
+  } else {
+    input_is_cmd_vel_ = false;
+  }
+
+  // Read Input References for servos
+  if (servo_input_reference_.size() == n_out_max) {
+    unsigned n_servos = 0;
+    for (unsigned i = 0; i < n_out_max; i++) {
+      if (!mavlink_interface_->IsInputMotorAtIndex(i)) {
+        servo_input_index_[n_servos++] = i;
+      }
+    }
+    servo_input_reference_.resize(n_servos);
+  }
+
+  for (int i = 0; i < servo_input_reference_.size(); i++) {
+    if (armed) {
+      servo_input_reference_[i] = actuator_controls[servo_input_index_[i]];
+    } else {
+      servo_input_reference_[i] = 0;
+    }
+  }
+
+  // Read Input References for motors
+  if (motor_input_reference_.size() == n_out_max) {
+    unsigned n_motors = 0;
+    for (unsigned i = 0; i < n_out_max; i++) {
+      if (mavlink_interface_->IsInputMotorAtIndex(i)) {
+        motor_input_index_[n_motors++] = i;
+      }
+    }
+    motor_input_reference_.resize(n_motors);
+  }
+
+  for (int i = 0; i < motor_input_reference_.size(); i++) {
     if (armed) {
-      input_reference_[i] = actuator_controls[input_index_[i]] * input_scaling_(i);
+      if (i == motor_input_reference_.size() - 1) {
+        // Set pusher motor velocity scaling if airframe is either fw or vtol. This assumes the pusher motor is the last motor!
+        double scaling;
+        if (servo_input_reference_.size() > 0 && servo_input_reference_.size() != n_out_max) {
+          scaling = static_cast<double>(fw_motor_vel_scaling_);
+        } else {
+          scaling = static_cast<double>(mc_motor_vel_scaling_);
+        }
+        motor_input_reference_[i] = actuator_controls[motor_input_index_[i]] * scaling;
+      }
+      else {
+        motor_input_reference_[i] = actuator_controls[motor_input_index_[i]] * static_cast<double>(mc_motor_vel_scaling_);
+      }
```

Comment:
I am not sure about the second condition `servo_input_reference_.size() != n_out_max`   
If we have n_out_max amount of servo, why we will use `mc_motor_vel_scaling_` for scailng the last servo? 

```suggestion
      double scaling = static_cast<double>(mc_motor_vel_scaling_);
      if (i == motor_input_reference_.size() - 1 && servo_input_reference_.size() > 0) {
        // Set pusher motor velocity scaling if airframe is either fw or vtol. This assumes the pusher motor is the last motor!
        scaling = static_cast<double>(fw_motor_vel_scaling_);
      }
     
      motor_input_reference_[i] = actuator_controls[motor_input_index_[i]] * scaling;
```

Reply:
I just adopted @jnippula solution. so we dont need to have this logic anymore